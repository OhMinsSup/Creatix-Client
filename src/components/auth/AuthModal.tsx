import React from 'react';
import styled, { css } from 'styled-components';
import { MdClose } from 'react-icons/md';
import transitions from '../../lib/styles/transitions';

const AuthModalBlock = styled.div<{ visible: boolean }>`
  display: flex;
  position: fixed;
  align-items: center;
  justify-content: center;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 990;
  background-color: rgba(228, 226, 226, 0.49);
  > .wrapper {
    width: 600px;
    height: 480px;
    max-width: none;
    padding: 0;
    margin: auto;
    outline: 0;
    overflow: hidden;
    transform-origin: bottom center;
    ${props =>
      props.visible
        ? css`
            animation: ${transitions.popInFromBottom} 0.4s forwards ease-in-out;
          `
        : css`
            animation: ${transitions.popOutToBottom} 0.2s forwards ease-in-out;
          `}

    .white-block {
      width: 100%;
      height: 100%;
      display: flex;
      flex-direction: column;
      justify-content: center;
      overflow: hidden;
      position: relative;
      background: #ffffff;
      background-repeat: no-repeat;
      border-radius: 4px;
      box-shadow: 0 2px 12px 0 rgba(0, 0, 0, 0.09);
      text-align: center !important;
      line-height: 20px;
      font-size: 16px;
      background-image: url(https://cdn-images-1.medium.com/max/900/1*gnUBeL-INIVDk0_GyL8x1g.png);
      background-size: 100%;
      .exit-wrapper {
        display: inline-block;
        position: absolute;
        top: 0;
        right: 0;
        outline: none;
        line-height: 1;
        padding: 15px;
        color: rgba(0, 0, 0, 0.54);
        font-size: 23px;
        svg {
          cursor: pointer;
        }
      }
    }
  }
`;

export interface AuthModalProps {
  visible: boolean;
  onAuthModalClose: () => void;
}
const AuthModal: React.SFC<AuthModalProps> = ({
  children,
  visible,
  onAuthModalClose,
}) => {
  const [closed, setClosed] = React.useState(true);
  React.useEffect(() => {
    let timeoutId: number | null = null;
    if (visible) {
      setClosed(false);
    } else {
      timeoutId = setTimeout(() => {
        setClosed(true);
      }, 200);
    }
    return () => {
      if (timeoutId) {
        clearTimeout(timeoutId);
      }
    };
  }, [visible]);

  if (!visible && closed) return null;
  return (
    <AuthModalBlock visible={visible}>
      <div className="wrapper">
        <div className="white-block">
          <div className="exit-wrapper">
            <MdClose onClick={onAuthModalClose} tabIndex={1} />
          </div>
          <div className="block-content">{children}</div>
        </div>
      </div>
    </AuthModalBlock>
  );
};

export default AuthModal;
