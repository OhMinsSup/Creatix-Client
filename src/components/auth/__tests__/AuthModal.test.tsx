import * as React from 'react';
import { render, fireEvent } from 'react-testing-library';
import AuthModal, { AuthModalProps } from '../AuthModal';

describe('AuthModal', () => {
  const setup = (props: Partial<AuthModalProps> = {}) => {
    const initialProps: AuthModalProps = {
      visible: false,
      onAuthModalClose: () => {},
    };
    return render(<AuthModal {...initialProps} {...props} />);
  };

  it('renders properly', () => {
    setup();
  });

  it('matches visible false snapshot', () => {
    const { container } = setup({
      visible: false,
    });
    expect(container).toMatchSnapshot();
  });

  it('matches visible true snapshot', () => {
    const { container } = setup({
      visible: true,
    });
    expect(container).toMatchSnapshot();
  });

  it('calls onToggleMode', () => {
    const onAuthModalClose = jest.fn();
    const { container } = setup({
      visible: true,
      onAuthModalClose,
    });

    const anchor = container.querySelector('svg');
    expect(anchor).toBeTruthy();
    if (!anchor) return;
    fireEvent.click(anchor);
    expect(onAuthModalClose.mock.calls.length).toBe(1);
  });
});
