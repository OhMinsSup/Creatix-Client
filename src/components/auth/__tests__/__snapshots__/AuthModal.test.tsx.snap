// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthModal matches visible false snapshot 1`] = `<div />`;

exports[`AuthModal matches visible true snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: fixed;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  z-index: 990;
  background-color: rgba(228,226,226,0.49);
}

.c0 > .wrapper {
  width: 600px;
  height: 480px;
  max-width: none;
  padding: 0;
  margin: auto;
  outline: 0;
  overflow: hidden;
  -webkit-transform-origin: bottom center;
  -ms-transform-origin: bottom center;
  transform-origin: bottom center;
  -webkit-animation: hwrkPK 0.4s forwards ease-in-out;
  animation: hwrkPK 0.4s forwards ease-in-out;
}

.c0 > .wrapper .white-block {
  width: 100%;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  overflow: hidden;
  position: relative;
  background: #ffffff;
  background-repeat: no-repeat;
  border-radius: 4px;
  box-shadow: 0 2px 12px 0 rgba(0,0,0,0.09);
  text-align: center !important;
  line-height: 20px;
  font-size: 16px;
  background-image: url(https://cdn-images-1.medium.com/max/900/1*gnUBeL-INIVDk0_GyL8x1g.png);
  background-size: 100%;
}

.c0 > .wrapper .white-block .exit-wrapper {
  display: inline-block;
  position: absolute;
  top: 0;
  right: 0;
  outline: none;
  line-height: 1;
  padding: 15px;
  color: rgba(0,0,0,0.54);
  font-size: 23px;
}

.c0 > .wrapper .white-block .exit-wrapper svg {
  cursor: pointer;
}

<div>
  <div
    class="c0"
  >
    <div
      class="wrapper"
    >
      <div
        class="white-block"
      >
        <div
          class="exit-wrapper"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            tabindex="1"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </div>
        <div
          class="block-content"
        />
      </div>
    </div>
  </div>
</div>
`;
